### ISSUE TITLE
Image source is clipped to -180,180 when terrain is enabled.

### ISSUE THREAD

**Issue by Hooterr** [maplibre/maplibre-gl-js#4088]
**maplibre-gl-js version**: 4.2.0

**browser**: chrome (not important)

### Steps to Trigger Behavior

 1. Add a style with terrain
 2. Add an image source that's outside -180,180 longitude
 4. Enable terrain 

### Link to Demonstration
https://jsbin.com/fiyisil/edit?html,output

```
const urlWms = "https://placehold.co/600x600/png";
const sourceName = "sourceName";
const wmsLayerName = "wms";
const imageCoordinates = [
    [-20, 60],
    [300, 60],
    [300, -60],
    [-20, -60]
];

map.addSource(sourceName, {
        'type': 'image',
        'url': urlWms,
        'coordinates': imageCoordinates
    });

map.addLayer({
    'id': wmsLayerName,
    'type': 'raster',
    'source': sourceName,
    "paint": {
        'raster-opacity': 0.33,
        'raster-fade-duration': 0
    }
});
```

Terrain on
![image](https://github.com/maplibre/maplibre-gl-js/assets/28758721/be4b8342-364b-43a7-8ff0-af287512c4ca)

Terrain off
![image](https://github.com/maplibre/maplibre-gl-js/assets/28758721/afa2b3f8-5673-49e6-87da-bc9158c26390)

### Expected Behavior
Image wraps correctly when terrain is enabled.

### Actual Behavior
Image seems to be clipped to -180,180 longitude
Also the opacity seems to be changing when terrain is enabled/disabled

**Comment by HarelM** [comment]
Is it possible that this is a duplicate of:
- #1559

**Comment by Hooterr** [comment]
Similar but different. I don't see problems with latitude. In the mentioned issue the problem seems to be with any combination of lat/lng, here the problem seems to be related to some sort of normalizing/clipping to a fixed range of -180,180.
Could be the same root cause though.

**Comment by Panthro22** [comment]
@Hooterr 
When the terrain is enabled, it is clipped to a specific tile. The image corner coordinates are then converted to that local tile's coordinate system so it will not go past the Tile edge as it is being cut off to fit that tile, even if it's near the border. If you zoom in to a street-level view, would the image disappear partially or entirely for you? I am asking as there is an issue of zooming in for an image's set near tile z-level of 1.

**Comment by Hooterr** [comment]
The image stays partially visible no matter what the zoom level is.

**Comment by nbald** [comment]
Same here

No terrain:
![Screenshot 2024-05-17 at 16-24-46 3D Terrain](https://github.com/maplibre/maplibre-gl-js/assets/1681443/66d951b9-b7ee-4164-a3c9-b1e2cfc6532b)ç

Terrain:
![Screenshot 2024-05-17 at 16-24-54 3D Terrain](https://github.com/maplibre/maplibre-gl-js/assets/1681443/5d146b81-7572-4ff2-9677-84ab59f945c7)

**Comment by nbald** [comment]
@wipfli could you please add a bounty label on this issue? My company gives a bounty of $200 USD for fixing this

**Comment by HarelM** [comment]
I've added a bounty label, It might be that solving #1559 will solve this issue as well, so there's a change for extra bounty for anyone that is interested.
Note that this issue's bounty will be paid by @nbald's company and not MapLibre.
CC: @ibesora, @sbachinin @neodescis @olsen232 - if any of you is interested in solving this.

**Comment by kleenkanteen** [comment]
@HarelM I'm interested in taking this on

**Comment by HarelM** [comment]
Great! I've signed this to you.

**Comment by sbachinin** [comment]
Hard to understand what this issue is about.
Originally it's about clipping to one globe. But the bounty is offered to fix the last example which is clipping to something else.
Two problems are not necessarily the same.
I think we should decide what is actually being fixed here

**Comment by HarelM** [comment]
To clarify: the 200$ bounty is on the problem @nbald is presenting in his post as his company will pay this bounty.
The other bounty, paid by maplibre, is on the linked issue and not this issue.
I only added the bounty tag here to indicate there's a bounty as per @nbald request.
I hope this is clearer now.
I might be that all these issues are the same root cause, but it might also not be the case.

**Comment by sbachinin** [comment]
I also doubt that the bounty problem is sufficiently described. Now there is only a screenshot and I think codepen would be great.
@kleenkanteen, can you reproduce the bug?

**Comment by nbald** [comment]
Here is my issue: https://jsbin.com/cahomamazu/edit?html,output
(same code as @Hooterr's, but with different coordinates)

```
        const imageCoordinates = [
            [-12, 55.4],
                [16, 55.4],
                [16, 37.5],
                [-12, 37.5]
        ];
```

Clipping occurs at longitude 0.

**Comment by terryluan12** [comment]
Hi, I noticed that kleenkanteen removed their assignment. Any chance I can give this issue a try?

**Comment by terryluan12** [comment]
Hi, I noticed that kleenkanteen removed their assignment. Any chance I can give this issue a try? @HarelM 

(@ibesora, @sbachinin @neodescis @olsen232 )

**Comment by HarelM** [comment]
Sure, feel free to give it a shot.

**Comment by terryluan12** [comment]
Hey! Just an update on this issue, I've narrowed it down to the `renderToTexture` function in `painter.ts` (and the `Painter` class). Currently parsing through the function, and attempting to translate it to what I know of OpenGL!

**Comment by HarelM** [comment]
@pstaszek please make sure to coordinate the work with @terryluan12 as I think both bugs are somewhat related...

**Comment by pstaszek** [comment]
Hey @terryluan12 
You might want to check what I described here: https://github.com/maplibre/maplibre-gl-js/issues/1559#issuecomment-2585400663 - I put there some initial outcomes from investigation, as well as some code that fixes (at least partially) the problem. We can work together on those the issues if you like - as they seems to be somehow related.

**Comment by geekdenz** [comment]
@nbald 
Are you still keen on rewarding the bounty on this?

@terryluan12 
You're assigned. Have you found anything like a commit where this bug is not there?

**Comment by pstaszek** [comment]
@HarelM @geekdenz 

I'll be happy to take over this issue - I recently fixed similar thing.

**Comment by HarelM** [comment]
Wasn't this fixed as part of
- #5598?

**Comment by pstaszek** [comment]
No, it wasn't - this problem still persists. I think there is different cause for this issue.

**Comment by HarelM** [comment]
Ok, assigning this to you as I haven't heard back from @terryluan12.

**Comment by HarelM** [comment]
@nbald is the bounty from your company still relevant?

**Comment by nbald** [comment]
> [@nbald](https://github.com/nbald) is the bounty from your company still relevant?

Sure, it is

**Comment by pstaszek** [comment]
Just want to let you know that I think I found root cause of the issue. It is related to wrapping. Below is the illustration:

![Image](https://github.com/user-attachments/assets/f13ab2f1-87fd-478e-b01f-48b925a7a683)

The image on tile with `wrap = 0` is rendered at some position, however on tile with `wrap = 1` it is rendered on the same position in relation to a tile as for previous one. It should be actually rendered with shift. This is the issue for terrain because tiles there are rendered to the texture, which then is rendered to the tile (images as well) so content can't overlap between tiles.

**Comment by HarelM** [comment]
@msbarry recently changed some wrapping related to query rendered features, not sure if this is related or not but you can check his recent PRs...

**Comment by pstaszek** [comment]
Thanks! However looks like his changes were unrelated.

You can find fixed code on this branch: https://github.com/pstaszek/maplibre-gl-js/tree/issue-4088-image-source-clipped-world-wrap-0

Please try it yourself, especially if you'll find still case that is not covered, let me now.

**Comment by pstaszek** [comment]
@nbald I also tried your example, it also seems fine after the change, but better try it yourself.

**Comment by pstaszek** [comment]
I have opened the PR with the fix: https://github.com/maplibre/maplibre-gl-js/pull/5989

**Comment by nbald** [comment]
Let me know when this issue is considered resolved, well tested and stable, so I can pay the bounty

**Comment by HarelM** [comment]
@nbald feel free to subscribe to the PR, it shouldn't be long until this will be in main and the released, I believe.

**Comment by nbald** [comment]
@pstaszek Our company offered a bounty for fixing this bug. I emailed you at the address listed on your public GitHub profile to arrange payment.
